oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/ExternalGroupPrincipalProvider.java

```diff
@@ -0,0 +1,353 @@
+            return new ExternalGroupPrincipal(principalName);
+        } else {
+            return null;
+        }
+    }
+
+    @Nonnull
+    @Override
+    public Set<Group> getGroupMembership(@Nonnull Principal principal) {
+        if (!(principal instanceof Group)) {
+            try {
+                if (principal instanceof ItemBasedPrincipal) {
+                    Tree t = root.getTree(((ItemBasedPrincipal) principal).getPath());
+                    return getGroupPrincipals(t);
+                } else {
+                    return getGroupPrincipals(userManager.getAuthorizable(principal));
+                }
+            } catch (RepositoryException e) {
+                log.debug(e.getMessage());
+            }
+
+        }
+        return ImmutableSet.of();
+    }
+
+    @Nonnull
+    @Override
+    public Set<? extends Principal> getPrincipals(@Nonnull String userID) {
+        try {
+            return getGroupPrincipals(userManager.getAuthorizable(userID));
+        } catch (RepositoryException e) {
+            log.debug(e.getMessage());
+            return ImmutableSet.of();
+        }
+    }
+
+    @Nonnull
+    @Override
+    public Iterator<? extends Principal> findPrincipals(@Nullable String nameHint, int searchType) {
+        if (PrincipalManager.SEARCH_TYPE_NOT_GROUP != searchType) {
+            Result result = findPrincipals(buildSearchPattern(nameHint), false);
+            if (result != null) {
+                return Iterators.filter(new GroupPrincipalIterator(nameHint, result), Predicates.notNull());
+            }
+        }
+
+        return Iterators.emptyIterator();
+    }
+
+    @Nonnull
+    @Override
+    public Iterator<? extends Principal> findPrincipals(int searchType) {
+        return findPrincipals(null, searchType);
+    }
+
+    //--------------------------------------------------------------------------
+    private Set<Group> getGroupPrincipals(@CheckForNull Authorizable authorizable) throws RepositoryException {
+        if (authorizable != null && !authorizable.isGroup()) {
+            Tree userTree = root.getTree(authorizable.getPath());
+            return getGroupPrincipals(userTree);
+        } else {
+            return ImmutableSet.of();
+        }
+    }
+
+    private Set<Group> getGroupPrincipals(@Nonnull Tree userTree) {
+        if (userTree.exists() && UserUtil.isType(userTree, AuthorizableType.USER) && userTree.hasProperty(REP_EXTERNAL_PRINCIPAL_NAMES)) {
+            PropertyState ps = userTree.getProperty(REP_EXTERNAL_PRINCIPAL_NAMES);
+            if (ps != null) {
+                Set<Group> groupPrincipals = Sets.newHashSet();
+                for (String principalName : ps.getValue(Type.STRINGS)) {
+                    groupPrincipals.add(new ExternalGroupPrincipal(principalName));
+                }
+                return groupPrincipals;
+            }
+        }
+        // no group principals defined
+        return ImmutableSet.of();
+    }
+
+    @CheckForNull
```

Comment:
maybe some explanation of what is exactly searched for. this helps understanding the code and counter-check if the query formulates what is intended.